File: benchmarks/16-interpolation/interpolation.m
(* Experiment 16 -- Interpolation.
   MEASURES src/interp.c and interp_mpfr.c: building an InterpolatingFunction and
   evaluating it, for both `Interpolation` (points -> function) and its value-only
   companion `ListInterpolation` (values on a regular grid -> function).  BUILD
   AND EVALUATE ARE SEPARATE ROWS on purpose -- a system can be fast to construct
   and slow to evaluate, and the two land in different files.

   EVERY CHECK EVALUATES AT A GRID NODE.  A node is a data point, so every
   interpolation method returns the stored value there regardless of the scheme
   (Wolfram's local Hermite vs scipy's global not-a-knot spline give different
   values BETWEEN nodes -- comparing them there would be a spurious CHECK-FAIL).
   Timing still measures the real construction / evaluation over the whole grid. *)

Get["../harness.m"];
Get["../data.m"];

(* InterpolationOrder is an OPTION, not a head -- `Names["InterpolationOrder"]`
   is {} even though `Interpolation[..., InterpolationOrder -> 1]` works, so it
   does not belong in the absence register. *)
require[{"Interpolation", "ListInterpolation", "InterpolatingFunction", "Fit"}];

(* MACHINE FLOATS, NOT EXACT RATIONALS.  `i/100` in the Wolfram language is an
   exact Rational, so `Interpolation` over such points keeps Sin[1/10] SYMBOLIC
   while the numpy column does float64 -- two different problems, and the first
   run of this experiment reported 70000x because of it, not because of any
   defect.  N[] pins both columns to machine doubles. *)
pts = N[Table[{i/10, Sin[i/10]}, {i, 1, 2000}]];

bench["Interpolation build, 2000 knots", Interpolation[pts];];
check["Interpolation build, 2000 knots", Length[pts]];

ifn = Interpolation[pts];
bench["Interpolation evaluate, 20000 points",
  Table[ifn[N[j/1000]], {j, 1000, 20999}];];
check["Interpolation evaluate, 20000 points", Round[10^6 N[ifn[2.5]]]];

(* Linear (order-1) build.  Keyed on nothing -- `Interpolation` is already
   required above, and InterpolationOrder is a supported option; the earlier
   benchIf gate on the option NAME silently SKIP'd this row on every run. *)
bench["Interpolation order 1 (linear) build",
  Interpolation[pts, InterpolationOrder -> 1];];
check["Interpolation order 1 (linear) build",
  Round[10^6 N[Interpolation[pts, InterpolationOrder -> 1][2.5]]]];

(* A 2-D grid: separable interpolation, a different code path. *)
grid = N[Table[Sin[i/10] Cos[j/10], {i, 1, 60}, {j, 1, 60}]];
benchIf["ListInterpolation 2-D 60x60", "ListInterpolation",
  ListInterpolation[grid];];
checkIf["ListInterpolation 2-D 60x60", "ListInterpolation",
  Round[10^6 N[ListInterpolation[grid][5., 5.]]]];

(* Evaluating on a packed array rather than element by element: this is the row
   that shows whether the interpolant has a vectorised path at all. *)
xs = rand01[{100000}] * 100 + 1;
bench["Interpolation over 10^5 array", ifn[xs];];
check["Interpolation over 10^5 array", Round[10^6 N[ifn[7.5]]]];

(* ==========================================================================
   ListInterpolation -- the value-only companion (src/interp.c).  A raw array of
   values on a regular grid, abscissae synthesised, delegating to the same
   Interpolation engine (so the same vectorised InterpolatingFunction object and
   the buffer-direct construction path off a packed value tensor).  The scipy
   analog is a CubicSpline over an arange grid (1-D), or RectBivariateSpline for
   the tensor case -- both node-exact interpolating cubics, so a grid-node check
   agrees to machine precision.
   ========================================================================== *)

(* 1-D build from 10^5 values on the integer grid 1, 2, ... .  The value tensor
   is auto-packed, so this exercises the buffer-direct construction path. *)
lvals = N[Table[Sin[i/10], {i, 1, 100000}]];
benchIf["ListInterpolation 1-D build, 10^5 values", "ListInterpolation",
  ListInterpolation[lvals];];
checkIf["ListInterpolation 1-D build, 10^5 values", "ListInterpolation",
  Round[10^6 N[ListInterpolation[lvals][25]]]];       (* node 25 -> Sin[2.5] *)

(* 1-D evaluate: the vectorised application path over a 10^5 packed query array,
   the ListInterpolation analog of the "Interpolation over 10^5 array" row. *)
lifn = ListInterpolation[lvals];
lxs = rand01[{100000}] * 99000 + 1;                    (* in-domain [1, 99001] *)
benchIf["ListInterpolation 1-D evaluate, 10^5 array", "ListInterpolation",
  lifn[lxs];];
checkIf["ListInterpolation 1-D evaluate, 10^5 array", "ListInterpolation",
  Round[10^6 N[lifn[7500]]]];                          (* node 7500 -> Sin[750] *)

(* 2-D build: a 200x200 separable grid, the tensor construction path at scale. *)
lgrid = N[Table[Sin[i/10] Cos[j/10], {i, 1, 200}, {j, 1, 200}]];
benchIf["ListInterpolation 2-D build, 200x200", "ListInterpolation",
  ListInterpolation[lgrid];];
checkIf["ListInterpolation 2-D build, 200x200", "ListInterpolation",
  Round[10^6 N[ListInterpolation[lgrid][50, 80]]]];    (* node -> Sin[5] Cos[8] *)

File: benchmarks/16-interpolation/interpolation.py
#!/usr/bin/env python3
"""Experiment 16 -- Interpolation (scipy.interpolate column).

Same kernels as ``interpolation.m``, same labels: the ``Interpolation`` rows,
then the value-only ``ListInterpolation`` companion (1-D build, 1-D vectorised
evaluate, 2-D build).

BUILD AND EVALUATE ARE SEPARATE ROWS on purpose: a system can be fast to
construct and slow to evaluate, and the two live in different files.

Mathematica's Interpolation defaults to order 3 (cubic), so CubicSpline is the
matching constructor; ListInterpolation over a value array is the same cubic on
a synthesised grid, so CubicSpline over an ``arange`` grid is its analog, and
RegularGridInterpolator(method="cubic") for the tensor case.  The two use
different end conditions (not-a-knot vs Wolfram's), so EVERY check evaluates at a
GRID NODE -- a data point, where every scheme returns the stored value -- rather
than between nodes, where the schemes legitimately disagree.
"""

import sys, os; sys.path.insert(0, os.path.dirname(os.path.dirname(os.path.abspath(__file__))))

import math

import numpy as np
from scipy.interpolate import CubicSpline, RectBivariateSpline, interp1d

from harness import bench, check, require, seed

# RectBivariateSpline, not RegularGridInterpolator(method="cubic"), is the tensor
# analog: both Mathilda's 2-D ListInterpolation and RectBivariateSpline are
# node-EXACT interpolating tensor cubics (RectBivariateSpline to 3e-16 at every
# node), so a grid-node check agrees.  RegularGridInterpolator's cubic method is
# a smoothing tensor spline that misses the data value at a node by up to ~4e-5,
# which silently CHECK-FAIL'd the 2-D row and discarded its timing.
require(["scipy.interpolate:CubicSpline", "scipy.interpolate:interp1d",
         "scipy.interpolate:RectBivariateSpline"])

xi = np.arange(1, 2001) / 10.0
yi = np.sin(xi)


def r6(x):
    return int(math.floor(float(x) * 1e6 + 0.5))


bench("Interpolation build, 2000 knots", lambda: CubicSpline(xi, yi))
check("Interpolation build, 2000 knots", len(xi))

cs = CubicSpline(xi, yi)
ev = np.arange(1000, 21000) / 1000.0
bench("Interpolation evaluate, 20000 points", lambda: cs(ev))
check("Interpolation evaluate, 20000 points", r6(cs(2.5)))

bench("Interpolation order 1 (linear) build",
      lambda: interp1d(xi, yi, kind="linear"))
check("Interpolation order 1 (linear) build",
      r6(interp1d(xi, yi, kind="linear")(2.5)))

gi = np.arange(1, 61) / 10.0
grid = np.outer(np.sin(gi), np.cos(gi))
ax = np.arange(1.0, 61.0)
bench("ListInterpolation 2-D 60x60",
      lambda: RectBivariateSpline(ax, ax, grid))
_rg = RectBivariateSpline(ax, ax, grid)
check("ListInterpolation 2-D 60x60", r6(_rg.ev(5.0, 5.0)))   # node -> sin(0.5)cos(0.5)

seed()
xs = np.random.random(100000) * 100.0 + 1.0
bench("Interpolation over 10^5 array", lambda: cs(xs))
check("Interpolation over 10^5 array", r6(cs(7.5)))

# -------------------------------------------------------------------------
# ListInterpolation -- values on a regular grid.  Mathilda synthesises the
# integer grid 1, 2, ... and delegates to the cubic Interpolation engine, so
# CubicSpline over np.arange(1, n+1) is the 1-D analog and
# RegularGridInterpolator(method="cubic") the tensor analog.
# -------------------------------------------------------------------------

# 1-D build from 10^5 values on the integer grid.
lxi = np.arange(1, 100001)
lvals = np.sin(lxi / 10.0)
bench("ListInterpolation 1-D build, 10^5 values", lambda: CubicSpline(lxi, lvals))
lcs = CubicSpline(lxi, lvals)
check("ListInterpolation 1-D build, 10^5 values", r6(lcs(25)))   # node 25 -> sin(2.5)

# 1-D vectorised evaluate over a 10^5 in-domain query array.
seed()
lxs = np.random.random(100000) * 99000.0 + 1.0
bench("ListInterpolation 1-D evaluate, 10^5 array", lambda: lcs(lxs))
check("ListInterpolation 1-D evaluate, 10^5 array", r6(lcs(7500)))  # node -> sin(750)

# 2-D build: a 200x200 separable grid.
lgi = np.arange(1, 201) / 10.0
lgrid = np.outer(np.sin(lgi), np.cos(lgi))
lax = np.arange(1.0, 201.0)
bench("ListInterpolation 2-D build, 200x200",
      lambda: RectBivariateSpline(lax, lax, lgrid))
_lrg = RectBivariateSpline(lax, lax, lgrid)
check("ListInterpolation 2-D build, 200x200", r6(_lrg.ev(50.0, 80.0)))  # node -> sin(5)cos(8)
